modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpDownloaderTaskExecutor.java

```diff
@@ -140,12 +198,41 @@ public void clusterChanged(ClusterChangedEvent event) {
             return;
         }
 
-        clusterService.removeListener(this);
-        if (ENABLED_SETTING.get(event.state().getMetadata().settings(), settings)) {
-            startTask(() -> clusterService.addListener(this));
-        } else {
-            stopTask(() -> clusterService.addListener(this));
+        if (taskIsBootstrapped.getAndSet(true) == false) {
+            this.atLeastOneGeoipProcessor = hasAtLeastOneGeoipProcessor(clusterService.state());
+            if (ENABLED_SETTING.get(event.state().getMetadata().settings(), settings)) {
+                startTask(() -> taskIsBootstrapped.set(false));
+            } else {
+                stopTask(() -> taskIsBootstrapped.set(false));
+            }
         }
+
+        if (event.metadataChanged() && event.changedCustomMetadataSet().contains(IngestMetadata.TYPE)) {
+            boolean newAtLeastOneGeoipProcessor = hasAtLeastOneGeoipProcessor(event.state());
```

Comment:
I don't think it will usually happen twice will it? It happens on bootstrap, and then happens below _only if_ the cluster state change includes an ingest metadata change (so only if someone modifies a pipeline definition), right?

Reply:
Ah yes, I hadn't thought about that. Looks fine to me then.